### DISCUSSION TITLE
The overrides system of Joomla is flawed

### DISCUSSION THREAD

**Discussion by dgrammatiko** [joomla/joomla-cms#38587] [2. Cold Features]
This comes in 2 parts: 
- CLI is not aware of the layouts (https://github.com/joomla/joomla-cms/pull/38585 tries to fix it)
- Both static assets and layouts overrides are checked per instance in every lifecycle (this has a huge perf/energy cost)

For the first I left a comment in @Fedik's PR
For the second: Templates are the only entity that HOLDS any overrides thus templates SHOULD provide a resolution (eg a json file with all the layout and asset overrides or even better a .php returning 2 objects one for the static files and one for the layouts, this will increase the efficiency if OPCache is enabled) instead of having to do a gazzilion of I/O file exists checks!

Since this is touching the architecture of the platform @laoneo @nibra @HLeithner might be for you

**Comment by Fedik** [comment]
> CLI is not aware of the layouts

Yeah, much of rendering (component view and layouts) thing is built for Web view, and overrides also. Quick fix for API was add "fake template":  https://github.com/joomla/joomla-cms/blob/3074a998b1bb42ee4778b563b8a2e52acf53a0ae/libraries/src/Application/ApiApplication.php#L194-L208

I probably would expect that for non Web context an override would look for a default site template. 
It would be more easy for end  User. 
Btw, Niels made RFC for "Decoupling Output" https://github.com/joomla/rfc/pull/36, I guess it is linked.

> Both static assets and layouts overrides are checked per instance in every lifecycle (this has a huge perf/energy cost)
> ... instead of having to do a gazzilion of I/O file exists checks!

You will be surprised how much "file exists" PHP doing under Class loaders.
But in general PHP do cache a loot of filesystem request.
It was "huge" when we used HDD, now it is fine :smile: 

UPD: In J4 css,js assets checked only once before actual rendering, in oposite to older joomla, when HTMLHelper was used for add these files.

**Reply by dgrammatiko** [reply]
> It was "huge" when we used HDD, now it is fine

Maybe you're right but for such cases only benchmarking the 2 versions can only give the actual state. 

Also something related looking at the future and when all of the browsers will support [import maps](https://caniuse.com/import-maps) having a JSON with all the overrides for JS files would be invaluable as the creation of an import map would be a simple object filter and json.stringify task. The project should always check where things are heading...

Also nice to the RFC, I know that @nibra already had a working prototype on this particular problem (and it's a nice one)
